### PR TITLE
drivers/periph/timer: add TIM_FLAG_SET_STOPPED flag

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

Currently it is not possible to initialize a timer with an alarm and then, at a later time, start the timer.
This is because it is assumed that the timer is always running, which leads to racy code like

```C
timer_set_periodic(timer, 0, timeout_us, TIM_FLAG_RESET_ON_MATCH);
timer_stop(timer);
```
which e.g. on sam0 currently does not what one would expect (the timer is not stopped).
Stopping the timer *before* setting the alarm will not work as many implementations assume the timer to be always running, thus when stopped internally in `timer_set_periodic()` they will restart it again.

Leverage that by adding a new flag to set the alarm but not run the timer. Implementations that do disable the timer for setting the alarm can then just skip the re-enabling step.
Implementations that do not stop the timer can just do so when the flag is set. 

### Testing procedure

`tests/periph_timer_periodic` has been extended to exercise the new flag. I tested the following implementations:

 - [x] sam0
 - [x] stm32
 - [x] atmega
 - [x] native
 - [x] nrf5x
 - [x] lpc23xx
 - [ ] gd32v
 - [ ] rpx0xx

### CODE REVIEW

File: drivers/dose/dose.c [hunk 1]
```diff
@@ -177,8 +177,8 @@ static void _dose_watchdog_cb(void *arg, int chan)
 static void _watchdog_init(unsigned timeout_us)
 {
     timer_init(DOSE_TIMER_DEV, US_PER_SEC, _dose_watchdog_cb, NULL);
-    timer_set_periodic(DOSE_TIMER_DEV, 0, timeout_us, TIM_FLAG_RESET_ON_MATCH);
-    timer_stop(DOSE_TIMER_DEV);
+    timer_set_periodic(DOSE_TIMER_DEV, 0, timeout_us,
+                       TIM_FLAG_RESET_ON_MATCH | TIM_FLAG_SET_STOPPED);
```
Comment (maribu) [inline drivers/dose/dose.c:181]: Care to sneak in the style fixes the CI suggests for this file in a separate commit?

**Review by maribu — COMMENTED** [review]
Code looks good to me. I have the hardware for the test at home and will do so later.

**Comment by maribu** [comment]
There is something wrong with the nRF5x implementation:

<details><summary>Test output on the <code>nrf52840dk</code></summary>

```
Help: Press s to start test, r to print it is ready
READY
s
START
main(): This is RIOT! (Version: 2022.04-devel-542-g27ffc2-ben)

Running Timer 0 at 1000000 Hz.
One counter cycle is 25000 ticks or 25 ms
Will print 'tick' every cycle.

TEST START
Running iteration 1 of 3
[0] tick
[0] tick
[0] tick
[0] tick
[0] tick
[0] tick
[0] tick
[0] tick
[0] tick
[0] tick
[0] tick
[0] tick

Cycles:
channel 0 = 12	[OK]
Running iteration 2 of 3
[0] tick
[0] tick
[0] tick
[0] tick
[0] tick
[0] tick
[0] tick
[0] tick
[0] tick
[0] tick
[0] tick
[0] tick

Cycles:
channel 0 = 12	[OK]
Running iteration 3 of 3
[0] tick
[0] tick
[0] tick
[0] tick
[0] tick
[0] tick
[0] tick
[0] tick
[0] tick
[0] tick
[0] tick
[0] tick

Cycles:
channel 0 = 12	[OK]
TTIM_FLAG_SET_STOPPED failed
TIM_FLAG_SET_STOPPED failed
TIM_FLAG_SET_STOPPED failed
TIM_FLAG_SET_STOPPED failed
TIM_FLAG_SET_STOPPED failed
TIM_FLAG_SET_STOPPED failed
TIM_FLAG_SET_STOPPED failed
TIM_FLAG_SET_STOPPED failed
TIM_FLAG_SET_STOPPED failed
TIM_FLAG_SET_STOPPED failed
TIM_FLAG_SET_STOPPED failed
TIM_FLAG_SET_STOPPED failed
TIM_FLAG_SET_STOPPED failed
TIM_FLAG_SET_STOPPED failed
TIM_FLAG_SET_STOPPED failed
TIM_FLAG_SET_STOPPED failed
TIM_FLAG_SET_STOPPED failed
TIM_FLAG_SET_STOPPED failed
TIM_FLAG_SET_STOPPED failed
TIM_FLAG_SET_STOPPED failed
TIM_FLAG_SET_STOPPED failed
TIM_FLAG_SET_STOPPED failed
TIM_FLAG_SET_STOPPED failed
TIM_FLAG_SET_STOPPED failed
TIM_FLAG_SET_STOPPED failed
TIM_FLAG_SET_STOPPED failed
TIM_FLAG_SET_STOPPED failed
TIM_FLAG_SET_STOPPED failed
TIM_FLAG_SET_STOPPED failed

...
```

</details>

File: cpu/gd32v/periph/timer.c [hunk 2]
```diff
@@ -164,6 +164,10 @@ int timer_set_periodic(tim_t tim, int channel, unsigned int value,
         return -1;
     }
 
+    if (flags & TIM_FLAG_SET_STOPPED) {
+        timer_stop(dev);
```
Comment (maribu) [inline cpu/gd32v/periph/timer.c:168]:
```suggestion
        timer_stop(tim);
```

**Comment by benpicco** [comment]
Should work now

<details><summary>Test output on the <code>nrf52840dk</code></summary>

```
2022-03-16 22:26:32,887 # START
2022-03-16 22:26:32,895 # main(): This is RIOT! (Version: 2022.04-devel-545-gcbcb89-periph_timer_periodic-set_stopped)
2022-03-16 22:26:32,895 # 
2022-03-16 22:26:32,898 # Running Timer 0 at 1000000 Hz.
2022-03-16 22:26:32,902 # One counter cycle is 25000 ticks or 25 ms
2022-03-16 22:26:32,905 # Will print 'tick' every cycle.
2022-03-16 22:26:32,905 # 
2022-03-16 22:26:32,906 # TEST START
2022-03-16 22:26:32,908 # Running iteration 1 of 3
2022-03-16 22:26:32,934 # [0] tick
2022-03-16 22:26:32,959 # [0] tick
2022-03-16 22:26:32,984 # [0] tick
2022-03-16 22:26:33,009 # [0] tick
2022-03-16 22:26:33,034 # [0] tick
2022-03-16 22:26:33,059 # [0] tick
2022-03-16 22:26:33,084 # [0] tick
2022-03-16 22:26:33,109 # [0] tick
2022-03-16 22:26:33,135 # [0] tick
2022-03-16 22:26:33,160 # [0] tick
2022-03-16 22:26:33,185 # [0] tick
2022-03-16 22:26:33,210 # [0] tick
2022-03-16 22:26:33,210 # 
2022-03-16 22:26:33,211 # Cycles:
2022-03-16 22:26:33,213 # channel 0 = 12	[OK]
2022-03-16 22:26:33,215 # Running iteration 2 of 3
2022-03-16 22:26:33,241 # [0] tick
2022-03-16 22:26:33,266 # [0] tick
2022-03-16 22:26:33,291 # [0] tick
2022-03-16 22:26:33,316 # [0] tick
2022-03-16 22:26:33,341 # [0] tick
2022-03-16 22:26:33,366 # [0] tick
2022-03-16 22:26:33,391 # [0] tick
2022-03-16 22:26:33,416 # [0] tick
2022-03-16 22:26:33,442 # [0] tick
2022-03-16 22:26:33,467 # [0] tick
2022-03-16 22:26:33,492 # [0] tick
2022-03-16 22:26:33,517 # [0] tick
2022-03-16 22:26:33,518 # 
2022-03-16 22:26:33,518 # Cycles:
2022-03-16 22:26:33,520 # channel 0 = 12	[OK]
2022-03-16 22:26:33,522 # Running iteration 3 of 3
2022-03-16 22:26:33,548 # [0] tick
2022-03-16 22:26:33,573 # [0] tick
2022-03-16 22:26:33,598 # [0] tick
2022-03-16 22:26:33,623 # [0] tick
2022-03-16 22:26:33,648 # [0] tick
2022-03-16 22:26:33,673 # [0] tick
2022-03-16 22:26:33,698 # [0] tick
2022-03-16 22:26:33,724 # [0] tick
2022-03-16 22:26:33,749 # [0] tick
2022-03-16 22:26:33,774 # [0] tick
2022-03-16 22:26:33,799 # [0] tick
2022-03-16 22:26:33,824 # [0] tick
2022-03-16 22:26:33,825 # 
2022-03-16 22:26:33,825 # Cycles:
2022-03-16 22:26:33,827 # channel 0 = 12	[OK]
2022-03-16 22:26:34,933 # TEST SUCCEEDED
```
</details>

Thank you for testing!

**Comment by benpicco** [comment]
Can I squash?

**Review by maribu — APPROVED** [review]
ACK. Looks good to me, the remaining untested changes are unscary enough.

**Pull request merged**